File: objectiveai-rs/src/auth/api_key.rs
//! API key types and definitions.

use crate::prefixed_uuid::PrefixedUuid;
use serde::{Deserialize, Serialize};

/// An ObjectiveAI API Key.
///
/// The format is always `apk` followed by 32 hexadecimal characters
/// representing a UUID (e.g., `apk1234567890abcdef1234567890abcdef`).
///
/// API keys are used to authenticate requests to the ObjectiveAI API.
pub type ApiKey = PrefixedUuid<'a', 'p', 'k'>;

/// An ObjectiveAI API Key with associated metadata.
///
/// This struct contains the API key itself along with information about
/// when it was created, when it expires (if ever), whether it has been
/// disabled, and user-provided name and description.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiKeyWithMetadata {
    /// The API key itself.
    pub api_key: ApiKey,
    /// The timestamp when the API key was created (RFC 3339 format).
    pub created: chrono::DateTime<chrono::Utc>,
    /// The timestamp when the API key expires, or `None` if it does not expire.
    pub expires: Option<chrono::DateTime<chrono::Utc>>,
    /// The timestamp when the API key was disabled, or `None` if it is active.
    pub disabled: Option<chrono::DateTime<chrono::Utc>>,
    /// The user-provided name of the API key.
    pub name: String,
    /// The user-provided description of the API key, or `None` if not provided.
    pub description: Option<String>,
}

File: objectiveai-js/src/auth/api_key/api_key.ts
import z from "zod";
import { convert, type JSONSchema } from "../../json_schema";

export const ApiKeySchema = z
  .string()
  .describe(
    "An ObjectiveAI API Key. The format is always `apk` followed by 32 hexadecimal characters.",
  );
export type ApiKey = z.infer<typeof ApiKeySchema>;
export const ApiKeyJsonSchema: JSONSchema = convert(ApiKeySchema);

export const ApiKeyWithMetadataSchema = z
  .object({
    api_key: ApiKeySchema,
    created: z
      .string()
      .describe("The RFC 3339 timestamp when the API key was created."),
    expires: z
      .string()
      .nullable()
      .describe(
        "The RFC 3339 timestamp when the API key expires, or null if it does not expire.",
      ),
    disabled: z
      .string()
      .nullable()
      .describe(
        "The RFC 3339 timestamp when the API key was disabled, or null if it is not disabled.",
      ),
    name: z.string().describe("The name of the API key."),
    description: z
      .string()
      .nullable()
      .describe(
        "The description of the API key, or null if no description was provided.",
      ),
  })
  .describe("An ObjectiveAI API Key with metadata.");
export type ApiKeyWithMetadata = z.infer<typeof ApiKeyWithMetadataSchema>;
export const ApiKeyWithMetadataJsonSchema: JSONSchema = convert(
  ApiKeyWithMetadataSchema,
);
